src/workers/parser/mapBindings.js

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */
+
+// @flow
+
+import { parseScript } from "./utils/ast";
+import { buildScopeList } from "./getScopes";
+import generate from "@babel/generator";
+import * as t from "@babel/types";
+import remove from "lodash/remove";
+
+import type { ColumnPosition } from "../../types";
+
+// translates new bindings `var a = 3` into `self.a = 3`
+// and existing bindings `var a = 3` into `a = 3` for re-assignments
+function globalizeAssignment(node, bindings) {
+  return node.declarations.map(node => {
+    const identifier = bindings.includes(node.id.name)
+      ? node.id
+      : t.memberExpression(t.identifier("self"), node.id);
+
+    return t.expressionStatement(
+      t.assignmentExpression("=", identifier, node.init)
+    );
+  });
+}
+
+function isTopLevel(ancestors) {
+  return ancestors.filter(ancestor => ancestor.key == "body").length == 1;
+}
+
+function replaceParent(ancestors, newNodes) {
+  const parent = ancestors[ancestors.length - 1];
+  const lastAncestor = ancestors[ancestors.length - 1];
+  const { index } = lastAncestor;
+  return parent.node[parent.key].splice(index, 1, ...newNodes);
+}
+
+function hasDestructuring(node) {
+  return node.declarations.some(declaration => t.isPattern(declaration.id));
+}
+
+function alreadyDefined(node, bindings) {
+  const names = node.declarations.map(dec => dec.id.name);
+  return names.some(name => bindings.includes(name));
+}
+
+export default function mapExpressionBindings(
+  expression: string,
+  bindings: string[] = []
+): string {
+  const ast = parseScript(expression);
+
+  t.traverse(ast, (node, ancestors) => {
+    const parent = ancestors[ancestors.length - 1];
+    if (
+      t.isVariableDeclaration(node) &&
+      isTopLevel(ancestors) &&
```

Comment:
true. i have a test that documents that

Reply:
Should we bail out if that's found? It seems fine to not handle all the cases, but since users have no insight into what we've decided to handle and what to ignore, it seems like partial-transformation will be pretty confusing.